### PR TITLE
SoC: SOF: ipc4-mtrace: handle bursts of log data directly

### PR DESCRIPTION
**Pull request by kv2019i**
Due to the limited shared memory used to transport logs, logs can easily get dropped if the firmware sends bursts of logs in small span of time.

One way to mitigate this is to just have firmware send IPCs of log availability at higher rate, at some overhead cost on firmware side.

This patch implements another way taking benefit of the higher processing capacity on host side. The mtrace client is modified to scheduled a delayed work item to monitor for more data to arrive, whenever firmware sends an IPC message about new logging data to be available. If no data is seen, the delayed work is cancelled and driver will wait until next IPC.

TODO:
 - clean out magic numbers (make tunable?)
 - check how much gap is caused by the trip to user-space (we don't update host read ptr until user-space has read the data out, which can hold up precious SRAM space between host core and FW)

Link: https://github.com/thesofproject/sof/issues/6669
Signed-off-by: Kai Vehmanen <kai.vehmanen@linux.intel.com>

### CODE REVIEW

File: sound/soc/sof/ipc4-mtrace.c [hunk 1]
```diff
@@ -309,6 +313,38 @@ static const struct file_operations sof_dfs_mtrace_fops = {
...
+static void sof_ipc4_mtrace_work(struct work_struct *work)
+{
+	struct sof_mtrace_core_data *core_data =
+		container_of(work, struct sof_mtrace_core_data, poller_work.work);
+	struct snd_sof_dev *sdev = core_data->sdev;
+	u32 new_ptr, old_ptr;
+
+	old_ptr = core_data->dsp_write_ptr;
+
+	/* TODO: use share helped function with sof_ipc4_mtrace_update_pos() */
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-mtrace.c:325]: typo: /helped/helper ?

File: sound/soc/sof/ipc4-mtrace.c [hunk 2]
```diff
@@ -309,6 +313,38 @@ static const struct file_operations sof_dfs_mtrace_fops = {
...
+
+	sof_mailbox_read(sdev, core_data->slot_offset + sizeof(u32),
+			 &new_ptr, 4);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-mtrace.c:328]: can fit to one line?

File: sound/soc/sof/ipc4-mtrace.c [hunk 3]
```diff
@@ -638,6 +678,10 @@ int sof_ipc4_mtrace_update_pos(struct snd_sof_dev *sdev, int core)
 
 	wake_up(&core_data->trace_sleep);
 
+	/* poll for a while in case DSP sends more data */
+	core_data->poller_idle = 0;
+	mod_delayed_work(system_wq, &core_data->poller_work, 1);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-mtrace.c:683]: This is going to run a 100 x 1 jiffies poll run even if the mtrace has not been opened by the user and the firmware happens to send messages?
Should this be done only if the mtrace is in use?

**Review by ujfalusi — COMMENTED** [review]
@kv2019i, as a debug feature I think this is good, but if I'm not mistaken it is going to have impact on systems where the mtrace is not actively used?
each position update will result a 100x loop, even if the trace is not actively used.

File: sound/soc/sof/ipc4-mtrace.c (hunk 3 above)
Comment (kv2019i) [reply sound/soc/sof/ipc4-mtrace.c:683]: Thanks @ujfalusi for early review. I think I'm going to shelve this for the time. It seems by modifying the rate limiting on FW side, we can avoid log loss with default settings at least. So I think this kernel PR would be more for special purpose debugging where you want to funnel a much large bandwidth of logs, and for some reason cannot use probes/DMA.

**Comment by plbossart** [comment]
@kv2019i what the direction here? Keep this PR alive or just close for now if we have already too many things to chase?

**Comment by kv2019i** [comment]
We can close this to keep up PRs clean, this is really a debugging tool.

**Pull request closed without merging**